File: samples_c/s124_graph_bfs.c
#include <stdio.h>

/* Girdi: ilk satirda v (dugum sayisi) e (kenar sayisi), ardindan e satir
   "a b" (yonsuz kenar a-b), sonra baslangic dugumu s.
   Cikti: BFS ziyaret sirasi (dugum numaralari, bosluk ile ayrilmis).
   Komsuluk listesi (adjacency list) sabit boyutlu dizilerle temsil edilir
   (basitlik icin dugum basina en fazla MAX_DEG komsu). */

#define MAX_V 100
#define MAX_DEG 20

static int adj[MAX_V][MAX_DEG];
static int adj_count[MAX_V];
static int visited[MAX_V];
static int queue[MAX_V];

static void add_edge(int a, int b) {
    adj[a][adj_count[a]++] = b;
    adj[b][adj_count[b]++] = a;
}

int main(void) {
    int v, e;
    if (scanf("%d %d", &v, &e) != 2) return 0;

    for (int i = 0; i < e; i++) {
        int a, b;
        if (scanf("%d %d", &a, &b) != 2) return 0;
        add_edge(a, b);
    }

    int start;
    if (scanf("%d", &start) != 1) return 0;

    int qhead = 0, qtail = 0;
    queue[qtail++] = start;
    visited[start] = 1;

    int first = 1;
    while (qhead < qtail) {
        int cur = queue[qhead++];
        if (!first) printf(" ");
        printf("%d", cur);
        first = 0;
        for (int i = 0; i < adj_count[cur]; i++) {
            int next = adj[cur][i];
            if (!visited[next]) {
                visited[next] = 1;
                queue[qtail++] = next;
            }
        }
    }
    printf("\n");
    (void)v;
    return 0;
}

File: translations_rust__gemini/s124_graph_bfs.rs
use std::io::{self, Read};

const MAX_V: usize = 100;
const MAX_DEG: usize = 20;

fn main() {
    let mut input = String::new();
    if io::stdin().read_to_string(&mut input).is_err() {
        return;
    }
    let mut tokens = input.split_whitespace();

    let _v: usize = match tokens.next().and_then(|s| s.parse().ok()) {
        Some(val) => val,
        None => return,
    };
    let e: usize = match tokens.next().and_then(|s| s.parse().ok()) {
        Some(val) => val,
        None => return,
    };

    let mut adj = [[0usize; MAX_DEG]; MAX_V];
    let mut adj_count = [0usize; MAX_V];

    for _ in 0..e {
        let a: usize = match tokens.next().and_then(|s| s.parse().ok()) {
            Some(val) => val,
            None => return,
        };
        let b: usize = match tokens.next().and_then(|s| s.parse().ok()) {
            Some(val) => val,
            None => return,
        };

        adj[a][adj_count[a]] = b;
        adj_count[a] += 1;
        adj[b][adj_count[b]] = a;
        adj_count[b] += 1;
    }

    let start: usize = match tokens.next().and_then(|s| s.parse().ok()) {
        Some(val) => val,
        None => return,
    };

    let mut visited = [0i32; MAX_V];
    let mut queue = [0usize; MAX_V];
    let mut qhead = 0;
    let mut qtail = 0;

    queue[qtail] = start;
    qtail += 1;
    visited[start] = 1;

    let mut first = true;
    while qhead < qtail {
        let cur = queue[qhead];
        qhead += 1;

        if !first {
            print!(" ");
        }
        print!("{}", cur);
        first = false;

        for i in 0..adj_count[cur] {
            let next = adj[cur][i];
            if visited[next] == 0 {
                visited[next] = 1;
                queue[qtail] = next;
                qtail += 1;
            }
        }
    }
    println!();
}
